File: app/src/main/java/com/example/myapplication/core/today/TodayResolver.kt
package com.example.myapplication.core.today

data class ScheduledWorkout(
    val id: String,
    val sequence: Int,
    val dueEpochDay: Long,
    val completed: Boolean,
)

sealed interface TodayResult {
    data object NoGoal : TodayResult
    data object GoalComplete : TodayResult
    data class Workout(val workout: ScheduledWorkout) : TodayResult
    data class Recovery(val nextDueEpochDay: Long) : TodayResult
}

object TodayResolver {
    fun resolve(
        hasActiveGoal: Boolean,
        workouts: List<ScheduledWorkout>,
        todayEpochDay: Long,
    ): TodayResult {
        if (!hasActiveGoal) return TodayResult.NoGoal

        val nextWorkout = workouts
            .asSequence()
            .filterNot(ScheduledWorkout::completed)
            .minWithOrNull(compareBy(ScheduledWorkout::sequence, ScheduledWorkout::dueEpochDay, ScheduledWorkout::id))
            ?: return TodayResult.GoalComplete

        return if (nextWorkout.dueEpochDay <= todayEpochDay) {
            TodayResult.Workout(nextWorkout)
        } else {
            TodayResult.Recovery(nextWorkout.dueEpochDay)
        }
    }
}

File: flutter/lib/core/today/today_resolver.dart
class ScheduledWorkout {
  final String id;
  final int sequence;
  final int dueEpochDay;
  final bool completed;

  ScheduledWorkout({
    required this.id,
    required this.sequence,
    required this.dueEpochDay,
    required this.completed,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ScheduledWorkout &&
          runtimeType == other.runtimeType &&
          id == other.id &&
          sequence == other.sequence &&
          dueEpochDay == other.dueEpochDay &&
          completed == other.completed;

  @override
  int get hashCode =>
      id.hashCode ^ sequence.hashCode ^ dueEpochDay.hashCode ^ completed.hashCode;
}

sealed class TodayResult {}

class TodayResultNoGoal extends TodayResult {
  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is TodayResultNoGoal && runtimeType == other.runtimeType;

  @override
  int get hashCode => 0;
}

class TodayResultGoalComplete extends TodayResult {
  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is TodayResultGoalComplete && runtimeType == other.runtimeType;

  @override
  int get hashCode => 0;
}

class TodayResultWorkout extends TodayResult {
  final ScheduledWorkout workout;
  TodayResultWorkout(this.workout);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is TodayResultWorkout &&
          runtimeType == other.runtimeType &&
          workout == other.workout;

  @override
  int get hashCode => workout.hashCode;
}

class TodayResultRecovery extends TodayResult {
  final int nextDueEpochDay;
  TodayResultRecovery(this.nextDueEpochDay);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is TodayResultRecovery &&
          runtimeType == other.runtimeType &&
          nextDueEpochDay == other.nextDueEpochDay;

  @override
  int get hashCode => nextDueEpochDay.hashCode;
}

class TodayResolver {
  static TodayResult resolve({
    required bool hasActiveGoal,
    required List<ScheduledWorkout> workouts,
    required int todayEpochDay,
  }) {
    if (!hasActiveGoal) return TodayResultNoGoal();

    final pendingWorkouts = workouts.where((w) => !w.completed).toList();
    if (pendingWorkouts.isEmpty) return TodayResultGoalComplete();

    pendingWorkouts.sort((a, b) {
      final seqCompare = a.sequence.compareTo(b.sequence);
      if (seqCompare != 0) return seqCompare;
      final dayCompare = a.dueEpochDay.compareTo(b.dueEpochDay);
      if (dayCompare != 0) return dayCompare;
      return a.id.compareTo(b.id);
    });

    final nextWorkout = pendingWorkouts.first;

    if (nextWorkout.dueEpochDay <= todayEpochDay) {
      return TodayResultWorkout(nextWorkout);
    } else {
      return TodayResultRecovery(nextWorkout.dueEpochDay);
    }
  }
}
